pkg/resource/resource.go

```diff
@@ -319,6 +319,19 @@ func ConnectionSecretMustBeControllableBy(u types.UID) ApplyOption {
 	}
 }
 
+// OverrideResourceVersion will override the resource version of the desired
+// object with that of the current. When used prior to a patch or update
+// operation it will force the operation without consideration of the resource
+// version that was set when the desired object was passed. It should only be
+// used in cases where it is desirable for the current object to be modified to
+// look exactly like the desired object.
+func OverrideResourceVersion() ApplyOption {
+	return func(_ context.Context, current, desired runtime.Object) error {
+		desired.(metav1.Object).SetResourceVersion(current.(metav1.Object).GetResourceVersion())
+		return nil
+	}
+}
+
```

Comment:
@negz @muvaf  I was testing this out and realized that now that we make the Update with the resource that we Get, we no longer need to override the resource version. I think we should update this to be called `ForceUpdate` and have it set `current = desired`, then set the resource version back to what is was in `current` before. Or maybe we just don't need this right now as I am currently mimicking the existing behavior in KA:

```
func (c *applicationResourceClient) sync(ctx context.Context, template *v1alpha1.KubernetesApplicationResource) (bool, error) {
	submitted := false

	// ApplyOptions are executed in the order that they are passed, so by the
	// time we reach the anonymous function we will already know that this KAR
	// is controllable by the KA.
	err := resource.NewAPIUpdatingApplicator(c.kube).Apply(ctx, template, resource.MustBeControllableBy(metav1.GetControllerOf(template).UID), func(_ context.Context, current, _ runtime.Object) error {
		c, ok := current.(*v1alpha1.KubernetesApplicationResource)
		if !ok {
			return errors.New("current object is not a KubernetesApplicationResource")
		}

		if c.Status.State == v1alpha1.KubernetesApplicationResourceStateSubmitted {
			submitted = true
		}

		c.SetLabels(template.GetLabels())
		c.SetAnnotations(template.GetAnnotations())
		c.Spec = *template.Spec.DeepCopy()

		// By the time we get here Apply will have already checked that this KAR
		// is controllable by the KA in the MustBeControllableBy ApplyOption, so
		// it is safe to overwrite the owner references with the template's.
		c.SetOwnerReferences(template.OwnerReferences)

		return nil
	})
	return submitted, errors.Wrapf(err, "cannot sync %s", v1alpha1.KubernetesApplicationResourceKind)
}
```

Reply:
Proposed:
```
// Apply changes to the supplied object. The object will be created if it does
// not exist, or updated if it does. If the object does exist and no
// ApplyOptions are passed, the update will be a no-op.
func (a *APIUpdatingApplicator) Apply(ctx context.Context, o runtime.Object, ao ...ApplyOption) error {
	m, ok := o.(metav1.Object)
	if !ok {
		return errors.New("cannot access object metadata")
	}

	if m.GetName() == "" && m.GetGenerateName() != "" {
		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
	}

	current := o.DeepCopyObject()

	err := a.client.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, current)
	if kerrors.IsNotFound(err) {
		// TODO(negz): Apply ApplyOptions here too?
		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
	}
	if err != nil {
		return errors.Wrap(err, "cannot get object")
	}

	for _, fn := range ao {
		if err := fn(ctx, current, o); err != nil {
			return err
		}
	}

	return errors.Wrap(a.client.Update(ctx, current), "cannot update object")
}
```

Reply:
The only different from the existing implementation is the last line:
`return errors.Wrap(a.client.Update(ctx, o), "cannot update object")`
becomes -->
`return errors.Wrap(a.client.Update(ctx, current), "cannot update object")`

Reply:
@hasheddan So if I follow correctly:

* If the supplied object does not exist we'll create it (without calling our apply options).
* If the supplied object does exist, the caller _must_ supply an `ApplyOption` that, given `current, desired runtime.Object` must transform `current` based on the state in `desired`.

This feels similar to the UX of `controllerutil.CreateOrUpdate`, but has a few nice properties. Firstly it satisfies our `Applicator` interface so we can more easily mock it out. Secondly it offers what feels like an easier to understand control flow than `CreateOrUpdate`. i.e. This proposed function requires the caller to:

1. Tell me what the desired object is. (The supplied object.)
1. Tell me how to transform a current object into the desired object.

Whereas `CreateOrUpdate` requires the caller to pass a function that tells it how to transform an object in any state into the desired object.

This makes the UX a little different from the patching applicator (which doesn't require an ApplyOption in the case the applied object already exists), but I think it's a good approach. 

I wonder if an `UpdateFn` would help with readability and brevity? I suspect we won't need `ctx` or to `return error` in most cases.

```go
o := &corev1.Pod{
    // Form the desired object.
}

_ := c.Apply(ctx, o, resource.UpdateFn(func(current, desired runtime.Object) {
    // Update current to match desired.
}))
```

I imagine the `UpdateFn` would be something like:

```go
func UpdateFn(fn func(current, desired runtime.Object)) ApplyOption {
    return func(_ context.Context, c, d runtime.Object) error {
        fn(c, d)
        return nil
    }
}
```

Reply:
@negz is the idea just to make it explicit that the body of the function is meant for the update operation? One could do this same thing with a normal `ApplyOption` and may choose to do additional modifications with additional `ApplyOptions`, which could be confusing. If we did introduce an `UpdateFn` I feel like it might be useful to call if after executing all `ApplyOptions` or something like that.

Reply:
> is the idea just to make it explicit that the body of the function is meant for the update operation?

Yes, and also to abbreviate the signature. I feel like this change will increase the places in which folks are supplying an anonymous function to `Apply` and I suspect in 99% of cases they won't need to supply a context or return an error. So:

```go
_ := c.Apply(ctx, o,
    resource.MustBeControllableBy(someUID),
    resource.UpdateFn(func(current, desired runtime.Object) {
        // Update current to match desired.
    }),
)
```

As opposed to:

```go
_ := c.Apply(ctx, o,
    resource.MustBeControllableBy(someUID),
    func(_ context.Context, current, desired runtime.Object) error {
        // Update current to match desired.
        return nil
    }),
)
```

I could be convinced that this doesn't add much value - just a thought.

Reply:
@negz You've sold me :) I believe this is ready to go now. Here is the `crossplane` usage of it for KA if helpful:

```
err := resource.NewAPIUpdatingApplicator(c.kube).Apply(ctx, template, resource.MustBeControllableBy(metav1.GetControllerOf(template).UID), resource.UpdateFn(func(current, desired runtime.Object) {
	c := current.(*v1alpha1.KubernetesApplicationResource)
	if c.Status.State == v1alpha1.KubernetesApplicationResourceStateSubmitted {
		submitted = true
	}

	c.SetLabels(template.GetLabels())
	c.SetAnnotations(template.GetAnnotations())
	c.Spec = *template.Spec.DeepCopy()

	// By the time we get here Apply will have already checked that this KAR
	// is controllable by the KA in the MustBeControllableBy ApplyOption, so
	// it is safe to overwrite the owner references with the template's.
	c.SetOwnerReferences(template.OwnerReferences)
}))
```